File: polyglot/DesignPatterns/Bridge/bridge.cpp
// Bridge design pattern
#include <iostream>
#include <memory>

// bridge implementation interface
class DrawAPI {
  public:
    virtual ~DrawAPI(){}
    virtual void drawCircle(int radius, int x, int y) const = 0;
};

// concrete bridge implementation classes implementing DrawAPI interface
class RedCircle : public DrawAPI {
  public:
    virtual ~RedCircle(){}
    void drawCircle(int radius, int x, int y) const override {
      std::cout << "Drawing Circle[ color: red  , radius: " 
        << radius << ", x: " << x << ", y: " << y << "]\n";
  }
};

class GreenCircle : public DrawAPI {
  public:
    virtual ~GreenCircle(){}
    void drawCircle(int radius, int x, int y) const override {
      std::cout << "Drawing Circle[ color: green, radius: " 
        << radius << ", x: " << x << ", y: " << y << "]\n";
  }
};

// create an abstract class Shape using the DrawAPI interface.
class Shape {
  protected:
    std::shared_ptr<DrawAPI> drawAPI;
    Shape(DrawAPI* drawAPI) : drawAPI(drawAPI){}
  public:
    virtual ~Shape(){}
    virtual void draw() const = 0;// abstraction need not follow implementation API
};

// create concrete class implementing the Shape interface (abstract class)
class Circle : public Shape {
  private:
    int x, y, radius;
  public:
    ~Circle(){}
    Circle(int x, int y, int radius, DrawAPI* drawAPI) : Shape(drawAPI){
    this->x = x;
    this->y = y;
    this->radius = radius;
  }
    void draw() const override {
    drawAPI->drawCircle(radius,x,y);
  }
};


// Use Shape and DrawAPI classes to draw different colored circles
int main(int argc, char* argv[], char* envp[]){
  // implementation of concrete circles selected at run time.
  typedef std::shared_ptr<Shape> ShapePtr;
  ShapePtr redCircle = ShapePtr(new Circle(100, 100, 10, new RedCircle()));
  ShapePtr greenCircle = ShapePtr(new Circle(100, 100, 10, new GreenCircle()));
  redCircle->draw();
  greenCircle->draw();
  return 0;
}

File: poly/DesignPatterns/Bridge/Bridge.java
// Bridge design pattern

// bridge implementation interface
interface DrawAPI {
  public void drawCircle(int radius, int x, int y);
}

// concrete bridge implementation classes implementing DrawAPI interface
class RedCircle implements DrawAPI {
  @Override
  public void drawCircle(int radius, int x, int y){
    System.out.println(
        "Drawing Circle[ color: red  , radius: "+radius+", x: "+x+ ", y: "+y+"]"
    );
  }
}

class GreenCircle implements DrawAPI {
  @Override
  public void drawCircle(int radius, int x, int y){
    System.out.println(
        "Drawing Circle[ color: green, radius: "+radius+", x: "+x+ ", y: "+y+"]"
    );
  }
}

// create an abstract class Shape using the DrawAPI interface.
abstract class Shape {
  protected DrawAPI drawAPI;
  protected Shape(DrawAPI drawAPI){this.drawAPI = drawAPI;}
  public abstract void draw();  // abstraction need not follow implementation API
}

// create concrete class implementing the Shape interface (abstract class)
class Circle extends Shape {
  private int x, y, radius;
  public Circle(int x, int y, int radius, DrawAPI drawAPI){
    super(drawAPI);
    this.x = x;
    this.y = y;
    this.radius = radius;
  }
  @Override
  public void draw(){
    drawAPI.drawCircle(radius,x,y);
  }
}

// Use Shape and DrawAPI classes to draw different colored circles
public class Bridge {
  public static void main(String[] args){
    // implementation of concrete circles selected at run time.
    Shape redCircle = new Circle(100, 100, 10, new RedCircle());
    Shape greenCircle = new Circle(100, 100, 10, new GreenCircle());
    redCircle.draw();
    greenCircle.draw();
  }
}
